### PR TITLE
f=srpAuth request handler

### PR DESCRIPTION
**Pull request by korDen**

### CODE REVIEW

File: SKELETON-KING/Program.cs [hunk 1]
```diff
@@ -1,13 +1,53 @@
 namespace SKELETON_KING;
 
+using Microsoft.EntityFrameworkCore;
+using System.Collections.Concurrent;
+using ZORGATH;
```
Comment (mrhappyasthma) [inline SKELETON-KING/Program.cs:5]: When you rebase this PR , you can now remove all local using directives. I added global Using.cs.

Comment (korDen) [reply SKELETON-KING/Program.cs:5]: SK doens't have Usings, right? Perhaps it doesn't even need one since it's just one file (and will probably stay that way).

Comment (mrhappyasthma) [reply SKELETON-KING/Program.cs]: Oh yeah, good point. I didn't realize this was in SK and not in Zorg.

File: ZORGATH/ChampionsOfNewerthRanks.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
+﻿namespace ZORGATH;
```
Comment (mrhappyasthma) [inline ZORGATH/ChampionsOfNewerthRanks.cs:1]: Can we pull the ChampionsOfNewerthRanks to its own PR? They even have tests from PLINKO we could port in PLINKO/Tests/Helpers/Stats/ChampionsOfNewerthRanksTests.cs

File: ZORGATH/PHP.cs [hunk 3]
```diff
@@ -60,15 +60,6 @@ private static ISerializer GetSerializerByType(Type type)
             if (keyType == typeof(string)) serializer = new HashSetStringSerializer();
```
Comment (mrhappyasthma) [inline ZORGATH/PHP.cs:60]: Are the PHP changes here necessitated by the SRP Auth response? It's a bit unclear to me the scope/impact of these.

If it is needed, can we add an explanation in the PR description?

File: ZORGATH/SrpAuthHandler.cs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    private static string? FinalizeSRPAuthentication(string login, SrpAuthSessionData autSessionData, string clientProof)
+    {
+        SrpParameters parameters = SrpParameters.Create<SHA256>(SrpAuthSessionData.N, SrpAuthSessionData.g);
+
+        // NOTE: HoN SRP requires a padded `g` (`parameters.Generator`) value for it's final `M`
+        // calculation. 
+        // The rfc5054 specification is unclear as to whether this 'should' be done or not. It
+        // is done in the Python SRP library:
+        // https://github.com/cocagne/pysrp/blob/4dfb111fffd671b7d97d803309fda2904e3ca1c8/srp/_pysrp.py#L205-L208
+        // It's unclear if a similar change can (or should) merge this into the C# SRP library.
```
Comment (mrhappyasthma) [inline ZORGATH/SrpAuthHandler.cs:207]: nit: They declined this change in the SRP lib and condoned our use of the workaround.

So perhaps let's change this comment to:

```
        // NOTE: HoN SRP requires a padded `g` (`parameters.Generator`) value for it's final `M`
        // calculation.  This is not handled by the SRP library since the rfc5054 specification is unclear as to 
        // whether this should be done. But it's required for HoN's SRP.
```

File: ZORGATH/SrpAuthHandler.cs [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+                /* tournamentGamesLost: */ 0,
+                /* tournamentTimesDisconnected: */ 0
+            ))
+            .FirstOrDefaultAsync();
+        if (accountDetails is null)
+        {
+            return new NotFoundObjectResult(PHP.Serialize(new AuthFailedResponse(AuthFailureReason.AccountNotFound)));
+        }
+
+        accountDetails = new(
```
Comment (mrhappyasthma) [inline ZORGATH/SrpAuthHandler.cs:86]: Why are we creating an account with hardcoded info here? Was this for local debugging? I believe it can be removed.

We only get to this line if `accountDetails` is not null (meaning it was pulled from the DB).

File: TRANSMUTANSTEIN/AuthResponseTest.cs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            {
+                new IgnoredListEntry(1, "DEVOURER")
+            }
+        };
+
+        string cookie = "C81C1CDA25B2F0E600681167BCC3D446";
+        string clientIpAddress = "127.0.0.1";
+        AuthResponse authResponse = new(accountDetails, cookie, clientIpAddress);
+
+        return;
```
Comment (mrhappyasthma) [inline TRANSMUTANSTEIN/AuthResponseTest.cs:148]: Seems like this test may be incomplete. It doesn't assert anything.

Comment (mrhappyasthma) [inline TRANSMUTANSTEIN/AuthResponseTest.cs:148]: nit: this return is not needed since it's the end of the method

File: ZORGATH/ChampionsOfNewerthRanks.cs (hunk 2 above)
Comment (korDen) [reply ZORGATH/ChampionsOfNewerthRanks.cs]: Done

File: ZORGATH/PHP.cs (hunk 3 above)
Comment (korDen) [reply ZORGATH/PHP.cs:64]: Yes, extracted into a separate PR.

File: ZORGATH/SrpAuthHandler.cs (hunk 4 above)
Comment (korDen) [reply ZORGATH/SrpAuthHandler.cs:119]: Done.

File: ZORGATH/SrpAuthHandler.cs (hunk 5 above)
Comment (korDen) [reply ZORGATH/SrpAuthHandler.cs]: Oops, yes.

File: TRANSMUTANSTEIN/AuthResponseTest.cs (hunk 6 above)
Comment (korDen) [reply TRANSMUTANSTEIN/AuthResponseTest.cs]: Yes, it was incomplete. Fixed now.

Comment (korDen) [reply TRANSMUTANSTEIN/AuthResponseTest.cs]: Fixed

File: ZORGATH/Account.cs [hunk 7]
```diff
@@ -4,16 +4,49 @@
...
+        SelectedUpgradeCodes = selectedUpgradeCodes ?? Array.Empty<string>();
+        AutoConnectChatChannels = autoConnectChatChannels ?? Array.Empty<string>();
+    }
+
     [Key]
     public int AccountId { get; set; }
 
     [Required]
-    [StringLength(20)]
+    [MaxLength(15)]
```
Comment (mrhappyasthma) [inline ZORGATH/Account.cs:22]: Do we enforce this on KONGOR account creation? Otherwise this may cause issues with backwards compatibility

File: ZORGATH/ElementUser.cs [hunk 8]
```diff
@@ -8,26 +8,67 @@
...
+
+    /// <summary>
+    ///     Number of gold coins that the user has.
+    /// </summary>
+    public int Points { get; set; }
+
+    /// <summary>
+    ///     Number of silver coins that the user has.
+    /// </summary>
+    public int MMPoints { get; set; }
```
Comment (mrhappyasthma) [inline ZORGATH/ElementUser.cs]: suggestion: if this won't break MerrickContext compatibility, let's use human understandable names here `GoldCoins` and `SilverCoins`. We can just update the comment to mention that HoN responses refer to these as `Points` and `MMPoints`.

File: ZORGATH/Friend.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace ZORGATH;
+
+public class Friend
+{
+    [Key]
+    // Unique key in the DB. Isn't used for anything else.
+    public int FriendId { get; set; }
+
+    // Id of out Account.
```
Comment (mrhappyasthma) [inline ZORGATH/Friend.cs:11]: nit: typo 'out' here. Is this ID of the friend or of the current user?

File: ZORGATH/SrpAuthHandler.cs [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        await bountyContext.Accounts
+            .Where(account => account.AccountId == accountDetails.AccountId)
+            .ExecuteUpdateAsync(update => update.SetProperty(account => account.Cookie, cookie));
+
+        string clientIpAddress = controllerContext.HttpContext.Connection.RemoteIpAddress.ToString();
+        long hostTime = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
+
+        // TODO: update these once we support custom account icons and the chatserver.
+        string chatServerUrl = "";
+        string icbUrl = "kongor.onlint";
```
Comment (mrhappyasthma) [inline ZORGATH/SrpAuthHandler.cs:93]: typo?

File: TRANSMUTANSTEIN/SrpAuthHandlerTest.cs [hunk 11]
```diff
@@ -0,0 +1,69 @@
+﻿namespace TRANSMUTANSTEIN;
+
+[TestClass]
+public class SrpAuthHandlerTest
```
Comment (mrhappyasthma) [inline TRANSMUTANSTEIN/SrpAuthHandlerTest.cs:4]: Can we also add a passing SRP test similar to `TestSrpAuthHandler` in KONGOR?

File: ZORGATH/Account.cs (hunk 7 above)
Comment (korDen) [reply ZORGATH/Account.cs:22]: KONGOR's max account name length is 14, there is however one account name that is longer than 14 which we should fix.

File: ZORGATH/Friend.cs (hunk 9 above)
Comment (korDen) [reply ZORGATH/Friend.cs:11]: Id of our account (current user).

File: ZORGATH/SrpAuthHandler.cs (hunk 10 above)
Comment (korDen) [reply ZORGATH/SrpAuthHandler.cs:93]: fixed

File: TRANSMUTANSTEIN/SrpAuthHandlerTest.cs (hunk 11 above)
Comment (korDen) [reply TRANSMUTANSTEIN/SrpAuthHandlerTest.cs:4]: There is one!

File: ZORGATH/Account.cs (hunk 7 above)
Comment (mrhappyasthma) [reply ZORGATH/Account.cs:22]: Should we set this to 14 then?

Comment (mrhappyasthma) [reply ZORGATH/Account.cs:22]: Or do we need a null character?

**Review by mrhappyasthma — APPROVED** [review]
